core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala

```diff
@@ -0,0 +1,303 @@
+   * by one node.
+   * If the pid/tag combo has no previous progress we start from 1 otherwise
+   * the old pidTagSequenceNr + 1.
+   * The future for getting the progress from C* is added to the lookup map
+   * so any further requests for that pid have to wait until the initial read
+   * is complete.
+   */
+  private def calculateTagPidSequenceNr(pid: String, bucket: TimeBucket): Future[Long] = {
+    // Have we got it already?
+    val tagPidSequenceNr: Future[SequenceNr] = pidTagSequenceNrs.get(pid) match {
+      case None =>
+        log.debug("First time seeing this pId {} since startup", pid)
+        session.selectTagProgress(pid).flatMap {
+          case None =>
+            log.debug("No previous progress for this tag/pid")
+            session.writeTagFirstBucket(pid, bucket).map(_ => 1)
+          case Some(t @ TagProgress(_, _, pidTagSequenceNr)) =>
+            log.debug("Tag progress for new pId: {}", t)
+            Future.successful(pidTagSequenceNr + 1)
+        }
+      case Some(n) =>
+        log.debug("Already cached. Previous tagPidSequenceNr: {}", n)
+        n.map(_ + 1)
+    }
+    pidTagSequenceNrs += (pid -> tagPidSequenceNr)
+    tagPidSequenceNr
+  }
+
+  /**
+   * Events should be ordered by sequence nr per pid
+   */
+  private def write(events: Vector[Serialized], remainingBuffer: Vector[Serialized]): Unit = {
+    // Work out all the pidTagSequenceNrs
+    val withPidTagSeqNr = events.map(s => {
+      calculateTagPidSequenceNr(s.persistenceId, s.timeBucket).map((s, _))
+    })
+
+    // Create a summary of this write for logging and to be written to the
+    // tag write progress table
+    val withSummary = Future.sequence(withPidTagSeqNr)
+      .map((wpt: Seq[(Serialized, SequenceNr)]) => {
+        val writeSummary = wpt.foldLeft(Map.empty[PersistenceId, PidProgress])((acc, next) => {
+          val (event, tagPidSequenceNr) = next
+          acc.get(event.persistenceId) match {
+            case Some(PidProgress(from, to, _, _)) =>
+              if (event.sequenceNr <= to)
+                throw new IllegalStateException(s"Expected events to be ordered by seqNr. ${event.persistenceId} Events: ${events.map(e => (e.persistenceId, e.sequenceNr))}")
+              acc + (event.persistenceId -> PidProgress(from, event.sequenceNr, tagPidSequenceNr, event.timeUuid))
+            case None =>
+              acc + (event.persistenceId -> PidProgress(event.sequenceNr, event.sequenceNr, tagPidSequenceNr, event.timeUuid))
+          }
+        })
+        (wpt, writeSummary)
+      })
+
+    val done = withSummary.flatMap {
+      case (writes, summary) =>
+        log.debug("Starting tag write. Summary: {}", summary)
+        session.writeBatch(writes).map(_ => TagWriteDone(summary))
+    }
+
+    // We never have concurrent writes for the same tag/node, go into writeInProgress state
+    context.become(writeInProgress(remainingBuffer))
+    done pipeTo self
```

Comment:
if the future failed `Status.Failure` message will be sent to `self`, but that is not taken care of

Reply:
All the c* errors should be dealt with now